File: kits/data-integration/providers/conflict_resolver/three_way_merge.ts
// Three-way merge conflict resolver — diffs both versions against a common ancestor
// Computes diff(ancestor, versionA) and diff(ancestor, versionB), merges non-overlapping
// changes cleanly. For text fields, attempts line-by-line three-way merge similar to git.

export interface Conflict {
  entityId: string;
  versionA: Record<string, unknown>;
  versionB: Record<string, unknown>;
  ancestor?: Record<string, unknown>;
  fieldConflicts: string[];
  timestampA?: number;
  timestampB?: number;
}

export interface ResolverConfig {
  options?: Record<string, unknown>;
}

export interface Resolution {
  winner: Record<string, unknown>;
  strategy: string;
  details: Record<string, unknown>;
}

interface FieldDiff {
  field: string;
  oldValue: unknown;
  newValue: unknown;
}

export const PROVIDER_ID = 'three_way_merge';
export const PLUGIN_TYPE = 'conflict_resolver';

function deepEqual(a: unknown, b: unknown): boolean {
  if (a === b) return true;
  if (a === null || b === null || typeof a !== 'object' || typeof b !== 'object') return false;
  const keysA = Object.keys(a as Record<string, unknown>);
  const keysB = Object.keys(b as Record<string, unknown>);
  if (keysA.length !== keysB.length) return false;
  for (const key of keysA) {
    if (!deepEqual((a as Record<string, unknown>)[key], (b as Record<string, unknown>)[key])) {
      return false;
    }
  }
  return true;
}

function computeDiff(ancestor: Record<string, unknown>, version: Record<string, unknown>): FieldDiff[] {
  const diffs: FieldDiff[] = [];
  const allKeys = new Set([...Object.keys(ancestor), ...Object.keys(version)]);

  for (const key of allKeys) {
    if (!deepEqual(ancestor[key], version[key])) {
      diffs.push({ field: key, oldValue: ancestor[key], newValue: version[key] });
    }
  }
  return diffs;
}

function threeWayTextMerge(ancestor: string, textA: string, textB: string): { merged: string; hasConflict: boolean } {
  const ancestorLines = ancestor.split('\n');
  const linesA = textA.split('\n');
  const linesB = textB.split('\n');
  const maxLen = Math.max(ancestorLines.length, linesA.length, linesB.length);

  const mergedLines: string[] = [];
  let hasConflict = false;

  for (let i = 0; i < maxLen; i++) {
    const orig = ancestorLines[i] ?? '';
    const lineA = linesA[i] ?? '';
    const lineB = linesB[i] ?? '';

    const aChanged = orig !== lineA;
    const bChanged = orig !== lineB;

    if (!aChanged && !bChanged) {
      mergedLines.push(orig);
    } else if (aChanged && !bChanged) {
      mergedLines.push(lineA);
    } else if (!aChanged && bChanged) {
      mergedLines.push(lineB);
    } else if (lineA === lineB) {
      mergedLines.push(lineA);
    } else {
      hasConflict = true;
      mergedLines.push(`<<<<<<< version_a`);
      mergedLines.push(lineA);
      mergedLines.push(`=======`);
      mergedLines.push(lineB);
      mergedLines.push(`>>>>>>> version_b`);
    }
  }

  return { merged: mergedLines.join('\n'), hasConflict };
}

export class ThreeWayMergeResolverProvider {
  resolve(conflict: Conflict, _config: ResolverConfig): Resolution {
    const ancestor = conflict.ancestor ?? {};
    const diffA = computeDiff(ancestor, conflict.versionA);
    const diffB = computeDiff(ancestor, conflict.versionB);

    const diffAFields = new Set(diffA.map(d => d.field));
    const diffBFields = new Set(diffB.map(d => d.field));

    const merged: Record<string, unknown> = { ...ancestor };
    const cleanMerges: string[] = [];
    const overlappingConflicts: string[] = [];
    const textMerges: string[] = [];

    // Apply non-overlapping diffs from A
    for (const diff of diffA) {
      if (!diffBFields.has(diff.field)) {
        merged[diff.field] = diff.newValue;
        cleanMerges.push(diff.field);
      }
    }

    // Apply non-overlapping diffs from B
    for (const diff of diffB) {
      if (!diffAFields.has(diff.field)) {
        merged[diff.field] = diff.newValue;
        cleanMerges.push(diff.field);
      }
    }

    // Handle overlapping changes
    for (const diff of diffA) {
      if (diffBFields.has(diff.field)) {
        const valA = conflict.versionA[diff.field];
        const valB = conflict.versionB[diff.field];

        if (deepEqual(valA, valB)) {
          merged[diff.field] = valA;
          cleanMerges.push(diff.field);
        } else if (typeof valA === 'string' && typeof valB === 'string') {
          const ancestorText = (ancestor[diff.field] as string) ?? '';
          const result = threeWayTextMerge(ancestorText, valA, valB);
          merged[diff.field] = result.merged;
          if (result.hasConflict) {
            overlappingConflicts.push(diff.field);
          } else {
            textMerges.push(diff.field);
          }
        } else {
          // Non-text overlapping conflict, default to version A
          merged[diff.field] = valA;
          overlappingConflicts.push(diff.field);
        }
      }
    }

    return {
      winner: merged,
      strategy: 'three_way_merge',
      details: {
        entityId: conflict.entityId,
        diffSetA: diffA.map(d => d.field),
        diffSetB: diffB.map(d => d.field),
        cleanMerges,
        textMerges,
        overlappingConflicts,
        hasAncestor: conflict.ancestor !== undefined,
        fullyResolved: overlappingConflicts.length === 0,
      },
    };
  }

  canAutoResolve(conflict: Conflict): boolean {
    const ancestor = conflict.ancestor ?? {};
    const diffA = computeDiff(ancestor, conflict.versionA);
    const diffB = computeDiff(ancestor, conflict.versionB);

    const diffAFields = new Set(diffA.map(d => d.field));

    for (const diff of diffB) {
      if (diffAFields.has(diff.field)) {
        const valA = conflict.versionA[diff.field];
        const valB = conflict.versionB[diff.field];
        if (!deepEqual(valA, valB)) {
          return false;
        }
      }
    }

    return true;
  }
}

export default ThreeWayMergeResolverProvider;

File: kits/data-integration/providers/conflict_resolver/three_way_merge.rs
// Three-way merge conflict resolver — diffs both versions against a common ancestor
// Computes diff(ancestor, versionA) and diff(ancestor, versionB), merges non-overlapping
// changes cleanly. For text fields, attempts line-by-line three-way merge similar to git.

use std::collections::{HashMap, HashSet};

pub const PROVIDER_ID: &str = "three_way_merge";
pub const PLUGIN_TYPE: &str = "conflict_resolver";

#[derive(Debug, Clone, PartialEq)]
pub enum Value {
    Null,
    String(String),
    Number(f64),
    Bool(bool),
    Array(Vec<Value>),
    Object(HashMap<String, Value>),
}

#[derive(Debug, Clone)]
pub struct Conflict {
    pub entity_id: String,
    pub version_a: HashMap<String, Value>,
    pub version_b: HashMap<String, Value>,
    pub ancestor: Option<HashMap<String, Value>>,
    pub field_conflicts: Vec<String>,
    pub timestamp_a: Option<u64>,
    pub timestamp_b: Option<u64>,
}

#[derive(Debug, Clone)]
pub struct ResolverConfig {
    pub options: HashMap<String, Value>,
}

#[derive(Debug, Clone)]
pub struct Resolution {
    pub winner: HashMap<String, Value>,
    pub strategy: String,
    pub details: HashMap<String, Value>,
}

#[derive(Debug)]
pub enum ResolverError {
    MergeFailure(String),
}

impl std::fmt::Display for ResolverError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            ResolverError::MergeFailure(msg) => write!(f, "Merge failure: {}", msg),
        }
    }
}

#[derive(Debug, Clone)]
struct FieldDiff {
    field: String,
    old_value: Option<Value>,
    new_value: Option<Value>,
}

fn compute_diff(ancestor: &HashMap<String, Value>, version: &HashMap<String, Value>) -> Vec<FieldDiff> {
    let mut diffs = Vec::new();
    let mut all_keys: HashSet<String> = HashSet::new();
    for key in ancestor.keys().chain(version.keys()) {
        all_keys.insert(key.clone());
    }

    for key in &all_keys {
        let old_val = ancestor.get(key);
        let new_val = version.get(key);
        if old_val != new_val {
            diffs.push(FieldDiff {
                field: key.clone(),
                old_value: old_val.cloned(),
                new_value: new_val.cloned(),
            });
        }
    }
    diffs
}

fn three_way_text_merge(ancestor: &str, text_a: &str, text_b: &str) -> (String, bool) {
    let ancestor_lines: Vec<&str> = ancestor.split('\n').collect();
    let lines_a: Vec<&str> = text_a.split('\n').collect();
    let lines_b: Vec<&str> = text_b.split('\n').collect();
    let max_len = ancestor_lines.len().max(lines_a.len()).max(lines_b.len());

    let mut merged_lines: Vec<String> = Vec::new();
    let mut has_conflict = false;

    for i in 0..max_len {
        let orig = ancestor_lines.get(i).copied().unwrap_or("");
        let line_a = lines_a.get(i).copied().unwrap_or("");
        let line_b = lines_b.get(i).copied().unwrap_or("");

        let a_changed = orig != line_a;
        let b_changed = orig != line_b;

        if !a_changed && !b_changed {
            merged_lines.push(orig.to_string());
        } else if a_changed && !b_changed {
            merged_lines.push(line_a.to_string());
        } else if !a_changed && b_changed {
            merged_lines.push(line_b.to_string());
        } else if line_a == line_b {
            merged_lines.push(line_a.to_string());
        } else {
            has_conflict = true;
            merged_lines.push("<<<<<<< version_a".to_string());
            merged_lines.push(line_a.to_string());
            merged_lines.push("=======".to_string());
            merged_lines.push(line_b.to_string());
            merged_lines.push(">>>>>>> version_b".to_string());
        }
    }

    (merged_lines.join("\n"), has_conflict)
}

pub struct ThreeWayMergeResolverProvider;

impl ThreeWayMergeResolverProvider {
    pub fn new() -> Self {
        Self
    }

    pub fn resolve(
        &self,
        conflict: &Conflict,
        _config: &ResolverConfig,
    ) -> Result<Resolution, ResolverError> {
        let empty = HashMap::new();
        let ancestor = conflict.ancestor.as_ref().unwrap_or(&empty);

        let diff_a = compute_diff(ancestor, &conflict.version_a);
        let diff_b = compute_diff(ancestor, &conflict.version_b);

        let diff_a_fields: HashSet<String> = diff_a.iter().map(|d| d.field.clone()).collect();
        let diff_b_fields: HashSet<String> = diff_b.iter().map(|d| d.field.clone()).collect();

        let mut merged = ancestor.clone();
        let mut clean_merges: Vec<String> = Vec::new();
        let mut overlapping_conflicts: Vec<String> = Vec::new();
        let mut text_merges: Vec<String> = Vec::new();

        // Apply non-overlapping diffs from A
        for diff in &diff_a {
            if !diff_b_fields.contains(&diff.field) {
                if let Some(val) = &diff.new_value {
                    merged.insert(diff.field.clone(), val.clone());
                } else {
                    merged.remove(&diff.field);
                }
                clean_merges.push(diff.field.clone());
            }
        }

        // Apply non-overlapping diffs from B
        for diff in &diff_b {
            if !diff_a_fields.contains(&diff.field) {
                if let Some(val) = &diff.new_value {
                    merged.insert(diff.field.clone(), val.clone());
                } else {
                    merged.remove(&diff.field);
                }
                clean_merges.push(diff.field.clone());
            }
        }

        // Handle overlapping changes
        for diff in &diff_a {
            if diff_b_fields.contains(&diff.field) {
                let val_a = conflict.version_a.get(&diff.field);
                let val_b = conflict.version_b.get(&diff.field);

                if val_a == val_b {
                    if let Some(v) = val_a {
                        merged.insert(diff.field.clone(), v.clone());
                    }
                    clean_merges.push(diff.field.clone());
                } else if let (Some(Value::String(sa)), Some(Value::String(sb))) = (val_a, val_b) {
                    let ancestor_text = match ancestor.get(&diff.field) {
                        Some(Value::String(s)) => s.as_str(),
                        _ => "",
                    };
                    let (merged_text, has_conflict) = three_way_text_merge(ancestor_text, sa, sb);
                    merged.insert(diff.field.clone(), Value::String(merged_text));
                    if has_conflict {
                        overlapping_conflicts.push(diff.field.clone());
                    } else {
                        text_merges.push(diff.field.clone());
                    }
                } else {
                    // Non-text overlapping conflict, default to version A
                    if let Some(v) = val_a {
                        merged.insert(diff.field.clone(), v.clone());
                    }
                    overlapping_conflicts.push(diff.field.clone());
                }
            }
        }

        let mut details = HashMap::new();
        details.insert("entity_id".to_string(), Value::String(conflict.entity_id.clone()));
        details.insert(
            "diff_set_a".to_string(),
            Value::Array(diff_a.iter().map(|d| Value::String(d.field.clone())).collect()),
        );
        details.insert(
            "diff_set_b".to_string(),
            Value::Array(diff_b.iter().map(|d| Value::String(d.field.clone())).collect()),
        );
        details.insert(
            "clean_merges".to_string(),
            Value::Array(clean_merges.iter().map(|f| Value::String(f.clone())).collect()),
        );
        details.insert(
            "text_merges".to_string(),
            Value::Array(text_merges.iter().map(|f| Value::String(f.clone())).collect()),
        );
        details.insert(
            "overlapping_conflicts".to_string(),
            Value::Array(overlapping_conflicts.iter().map(|f| Value::String(f.clone())).collect()),
        );
        details.insert(
            "has_ancestor".to_string(),
            Value::Bool(conflict.ancestor.is_some()),
        );
        details.insert(
            "fully_resolved".to_string(),
            Value::Bool(overlapping_conflicts.is_empty()),
        );

        Ok(Resolution {
            winner: merged,
            strategy: "three_way_merge".to_string(),
            details,
        })
    }

    pub fn can_auto_resolve(&self, conflict: &Conflict) -> bool {
        let empty = HashMap::new();
        let ancestor = conflict.ancestor.as_ref().unwrap_or(&empty);

        let diff_a = compute_diff(ancestor, &conflict.version_a);
        let diff_b = compute_diff(ancestor, &conflict.version_b);

        let diff_a_fields: HashSet<String> = diff_a.iter().map(|d| d.field.clone()).collect();

        for diff in &diff_b {
            if diff_a_fields.contains(&diff.field) {
                let val_a = conflict.version_a.get(&diff.field);
                let val_b = conflict.version_b.get(&diff.field);
                if val_a != val_b {
                    return false;
                }
            }
        }

        true
    }
}
